java/src/main/native/src/AggregationJni.cpp

```diff
@@ -307,8 +307,8 @@ JNIEXPORT jlong JNICALL Java_ai_rapids_cudf_Aggregation_createHostUDFAgg(JNIEnv*
   JNI_NULL_CHECK(env, udf_native_handle, "udf_native_handle is null", 0);
   try {
     cudf::jni::auto_set_device(env);
-    auto const udf_ptr = reinterpret_cast<cudf::host_udf_base const*>(udf_native_handle);
-    auto output        = cudf::make_host_udf_aggregation(udf_ptr->clone());
```

Comment:
What is wrong with this code? We would clone the underlying host_udf to be used, which then let the java caller be free to close the HostUDFWrapper whenever they wanted to.

I am also fine with the `host_udf_base*` being created on demand like with the `createUDFInstance()` API and passed down to this code. But in that case we should not need to modify the C++ API to make is less secure. Instead we should wrap it in a unique_ptr and pass that to `cudf::make_host_udf_aggregation` so the aggregation itself can delete it.

Reply:
This function `Java_ai_rapids_cudf_Aggregation_createHostUDFAgg` is only called by `HostUDFAggregation`, in function:
```java
        long createNativeInstance() {
            // The created host Agg instance takes the ownership of the passed in UDF instance.
            // When a host Agg instance is released, the UDF instance in it is also released.
            return Aggregation.createHostUDFAgg(wrapper.createUDFInstance());
        }
```

Refer to the above code, the passed in udf_native_handle is created by `wrapper.createUDFInstance`,  if document the `createUDFInstance`  as private function, then it's secure. And the new version directly use the passed in UDF instance, it's more effective without clone.

Reply:
I see no reason why so we only create the native UDF instance only on demand but not upon initialization because the native instance will be used eventually, isn't it? Or did I missed something here?

Reply:
> only create the native UDF instance only on demand 

This is reasonable, in this PR https://github.com/NVIDIA/spark-rapids/pull/11638, I have to create a UDF instance, and pass into GroupByAggregation, and have another place to close the instance in `HyperLogLogPlusPlusHostUDF`

Below code is memory leaking hll.
```java
  override lazy val groupByAggregate: GroupByAggregation = {
      val hll =new HyperLogLogPlusPlusHostUDF(AggregationType.GroupBy, precision)
      GroupByAggregation.hostUDF(hll)
    }
```

If we create the UDF instance lazily, and bind the UDF instance to a Aggregation instance like this PR, then we are free to release the UDF instance.